README.md
# a.nwesen.de: Ein Dienst für Anwesenheitslisten für Hochschulen

Lutz Prechelt, 2021-11-19

https://github.com/prechelt/anwesende

Simple attendance registration for universities having pandemics.
"anwesende" is German for "people that are being present".


**DE**: Eine sehr einfache Lösung für die Verfolgung von Infektionsketten,
die stark auf menschliche Urteilskraft und manuelle Schritte baut, 
um die Software einfach und den Gebrauch flexibel zu halten.

**EN**: A very simple web application to help tracing infection chains 
at universities.
Makes heavy use of human judgment and manual operations 
to keep the software simple and its use flexible.  
Designed for universities in the German-language area, so parts
of the application are German+English, parts are German-language only.
The code and technical documentation (Sections 4 and 5) are in English.

# 1. Abläufe / Business processes (German only)

## 1.1. Überblick

Rollen: 
- Dienst `a.nwesen.de` (kurz: Dienst): 
  Die hier web-basierte Funktionalität,
  die die hier beschriebene Software entfaltet, wenn man sie betreibt. 
- Teilnehmende Hochschule (kurz: Hochschule): 
  Eine Hochschule, die den Dienst betreibt oder mit nutzt.
- Besucher/in: Eine Person, deren Anwesenheit in einem Raum der Hochschule
  erfasst werden soll. 
- Datenverwalter/in: Die Person, die Lesezugriff auf die erfassten
  Anwesenheitsdaten hat und anderen Berechtigten selektiv Zugang verschafft.

Schritte des Gesamtablaufs:
1. Eine Einheit einer teilnehmenden Hochschule übermittelt der Datenverwalter/in 
   eine Liste von Räumen und Sitzplätzen, 
   siehe `Räume übermitteln` unten.
2. Sie erhält im Gegenzug eine PDF-Datei mit einem QR-Code für
   jeden Sitzplatz
   und klebt den QR-Code dauerhaft am zugehörigen Sitzplatz auf.
3. Ein/e Besucher/in scannt den QR-Code an ihrem Sitzplatz
   und gibt minimale Daten für die Verfolgung ein;
   siehe `Besucher/innen/sicht` unten.
4. Der Dienst speichert die Platzdaten, Personendaten und den Zeitpunkt.
5. Im Infektionsfall ruft die Datenverwalter/in die Anwesenheitsdaten für die
   Kontaktgruppen einer infizierten Person im passenden Zeitfenster 
   (z.B. 3 Tage) ab,
   um sie dem Gesundheitsamt zu übermitteln;
   siehe `Anwesenheitsdaten abrufen` unten.

Große Teile dieser Beschreibung sind auch in der Software selbst enthalten:
Der Überblick in 
[anwesende/templates/room/home.html](anwesende/templates/room/home.html)
([Homepage](https://anwesende.imp.fu-berlin.de/)),
die Beschreibung der Schritte 1, 3 und 5 in 
[anwesende/templates/room/import.html](anwesende/templates/room/import.html)
([Raum-Import](https://anwesende.imp.fu-berlin.de/import)),
[anwesende/templates/room/privacy.html](anwesende/templates/room/privacy.html) und
[anwesende/templates/room/search.html](anwesende/templates/room/search.html).
([Datenabfrage](https://anwesende.imp.fu-berlin.de/search)).

Weitere Informationen finden sich in den FAQs unter
[anwesende/templates/room/faq.html](anwesende/templates/room/faq.html)).
([FAQs](https://anwesende.imp.fu-berlin.de/faq)),


## 1.2 Besucher/innen/sicht

Rollen: 
- Dienst `a.nwesen.de`, 
- Besucher/in ('ich') einer teilnehmenden Hochschule.

1. Ich komme in einen offenen Raum der teilnehmenden Hochschule.
   An meinem Sitzplatz klebt ein QR-Code. Ich scanne ihn mit meinem
   Smartphone und lande in meinem Webbrowser auf einer Seite mit
   ungefähr so einem Namen: `http://a.nwesen.de/1234567890`.
2. Der Dienst zeigt mir Datenschutzhinweise.
3. Ich gebe meine Daten ein (Vorname, Name, Mobilfunknummer, Email,
   Straße/Hausnummer, PLZ, Ort)
   und sende das Formular ab.  
   Ab dem zweiten Aufruf ist das Formular wegen eines Cookies
   sogar schon vor-ausgefüllt und ich muss es nur absenden.

Anmerkungen: 
- Der Dienst registriert nur das **Kommen** in einen Raum,
  nicht das **Verlassen** (in der Annahme, dass das nicht gut klappen würde,
  aber die betroffenen Zeiträume trotzdem meist gut eingrenzbar sind).
- Zum Scannen von QR-Codes gibt es zahllose Apps.
  Einfach im PlayStore/AppStore nach `QR Code` suchen.


## 1.3 Räume übermitteln

Rollen: 
Mitarbeiter/in einer teilnehmenden Hochschule,
Datenverwalter/in.

1. Die Mitarbeiter/in lädt sich die 
   [Excel-Vorlage zur Raumübermittlung](anwesende/static/xlsx/roomdata-example.xlsx)
   und füllt sie aus:
   - Die Überschriftzeile und die Zellenformate keinesfalls ändern!
   - Jeder Raum bekommt eine Zeile.
   - Für alle Einträge bitte genau die nachfolgenden Hinweise beachten:
   - `organization`: Domainname der teilnehmenden Hochschule, z.B.
     `fu-berlin.de` für die Freie Universität Berlin.  
     Dieser Wert ist in allen Zeilen der Datei gleich.
   - `department`: Intern gebräuchlicher (Kurz)Name der Hochschuleinheit, z.B.
     `MathInf` für den Fachbereich Mathematik und Informatik.  
     Auch dieser Wert ist in der Regel in allen Zeilen der Datei gleich,
     weil die Mitarbeiter/in nur zu diesem einem Bereich gehört.
   - `building`: Gebäudebezeichnung, meist Straße und Hausnummer, z.B.
     `Takustr. 9` für das Informatik-Hauptgebäude in der Takustraße 9.  
     Diesen Wert sollte man (genau wie alle anderen) am besten kurz halten,
     da er auf dem QR-Code-Schild mit ausgedruckt wird und dort
     die verfügbare Breite nicht überschreiten sollte.  
     Der Name muss nicht unbedingt exakt zum Straßenschild passen.
     Wenn es also von einem langen Straßennamen eine in dieser
     Hochschuleinheit allgemein gebräuchliche Abkürzung gibt,
     könnte man auch die verwenden.
   - `room`: Raumbezeichnung, in der Regel eine Raumnummer, z.B.
     `055` für den Seminarraum 055 im Erdgeschoss.  
     Dies sollte zur Beschriftung an der Eingangstür des Raums passen.
   - `row_dist`: minimaler Abstand der Sitzreihen in Metern.
   - `seat_dist`: minimaler Abstand zweiter benachbarter Sitze (links/rechts) in Metern.
   - `seat_last`: Letzter Sitz, z.B. `r2s7` für Reihe 2, Sitz 7 in einem
     Raum mit 14 Sitzen in zwei Reihen. Der erste Sitz ist immer `r1s1`.  
     Diese Sitznamen werden auf den QR-Code-Schildern
     gut lesbar mit ausgedruckt.
     Wenn Reihen unterschiedlich viele Sitze haben, muss die maximale Anzahl
     auch für die letzte Reihe angegeben werden (selbst wenn die in Wirklichkeit
     kürzer sein sollte) und manche QR-Codes werden dann
     nicht mit aufgeklebt; hier ein 
     [Beispiel mit `seat_last` r5s4](anwesende/static/images/seatname-example.png).  
     Die Software nimmt je 1,4m Sitzabstand an und gibt 
     im Schritt "Anwesenheitsdaten abrufen" resultierende Abstände mit aus.
2. Die Mitarbeiter/in schickt die Excel-Datei per Email an
   die Datenverwalter/in.
3. Die Datenverwalter/in prüft die Daten auf Plausibilität,
   liest die Datei in den Dienst ein
   und erhält eine Seite mit QR-Codes als Ergebnis.
   Sie enthält für jeden Sitzplatz einen QR-Code mit beschreibender Beschriftung
   wie es in der Excel-Datei deklariert war.
5. Die Datenverwalter/in druckt diese Webseite in eine PDF-Datei
   und schickt sie der Mitarbeiter/in per Email.  
6. Die Mitarbeiter/in druckt die QR-Codes aus und klebt jeden davon
   auf den entsprechenden Sitzplatz im entsprechenden Raum.  
   Der QR-Code sollte vollständig mit mattem transparentem Klebeband
   bedeckt sein, damit er lange lesbar bleibt.

Varianten:

Schritt 3b: Ist die Datei fehlerhaft und lässt sich nicht einlesen,
korrigiert die Datenverwalter/in offensichtliche Fehler selbst
und klärt andernfalls die Korrekturen mit der Mitarbeiter/in.

Anmerkungen:

- Obige Beschreibung gilt für den Bereich der Lehre. 
  Im Bürobereich wird hingegen grober erfasst:
  Ein Bereich, in dem sich Beschäftigte vermutlich gelegentlich auch 
  länger begegnen (Korridor, Küche, Besprechungsraum, Bürobesuche), 
  wird zu einem "Raum" zusammengefasst, z.B. ein ganzer Flur.  
  Dadurch muss jede/r nur einmal pro Tag eine Eingabe machen und die 
  Kontaktgruppen fallen eher zu groß aus als zu klein. 
  Das ist deshalb sinnvoll, weil es wesentlich einfacher ist, 
  eine zu lange Liste möglicher Kontakte zu bereinigen, 
  als fehlende Kontakte aufzuspüren.
- Innerhalb eines solchen "Raums" kann man dann Räume (oder
  Teile davon) als Sitzplätze interpretieren und die
  Beschäftigten anhalten, sich ggf. mehrfach anzumelden,
  wenn sie länger an einem anderen "Sitzplatz" (z.B. der Küche)
  als ihrem üblichen eigenen sind,
  damit bei der Nachverfolgung das Sortieren in Kontakte
  und Nichtkontakte von den Anmeldedaten besser unterstützt wird.  
- Beispiel: A und B arbeiten gleichzeitig am "Sitzplatz" 1 
  (tatsächlich Raum 33).
  C, D, E, F, G, H arbeiten zur selben Zeit an den "Sitzplätzen" 
  2, 2, 3, 4, 5, 5, 
  (tatsächlich den Räumen 34, 34, 35, 36, 37, 37).  
  Die Küche ist als "Sitzplatz" 7 registriert.  
  Wenn sich A infiziert, ist sofort B als Kontakt klar.
  Aber was ist mit B bis H?  
  A, C und F haben sich am fraglichen Tag auch mal am Sitzplatz
  7 angemeldet. C zu einer anderen Zeit als A, aber F zu einer überlappenden.   
  Bei der Abfrage würden also B bis H alle in der Kontakteliste
  aufgeführt, aber B und F sind als wahrscheinlichste Kontakte
  leicht erkennbar, denn B hat den gleichen "Sitzplatz" 1 wie A
  und F hat den gleichen "Sitzplatz" 7 wie A.
  Den hat auch C, wird aber mangels Zeitüberlappung nur auf der
  gleichen mäßigen Risikostufe geführt wie die übrigen.


## 1.4 Anwesenheitsdaten abrufen

Rollen:
Infiziertes Mitglied der Hochschule,
Mitarbeiter/in der Hochschuleinheit,
Datenverwalter/in

1. Ist ein Mitglied (oder ein/e Besucher/in) der Hochschule infiziert, 
   so informiert es die betreffenden Hochschuleinheit(en) darüber,
   an welchen Tagen eine Infektiösität bestanden haben kann.
2. Die Mitarbeiter/in der Hochschuleinheit, die bei der Datenverwalter/in
   als abrufberechtigt bekannt ist, übermittelt diesen Zeitraum und
   die Personendaten (Name, Telefon, Email) der infizierten Person per
   Email an die Datenverwalterin.  
3. Die Datenverwalter/in ruft die zugehörigen Kontaktpersonen ab und übermittelt
   sie an die Mitarbeiter/in.
   Als Kontaktpersonen gelten alle, die Einträge im selben Raum haben,
   die sich zeitlich mindestens 15 Minuten mit der Anwesenheit der
   infizierten Person überlappen.
4. Die Mitarbeiter/in bereinigt die Daten per Augenschein und Rückfragen 
   und entfernt ggf. überzählige Einträge.
5. Die Mitarbeiter/in übermittelt die bereinigten Daten dem
   Gesundheitsamt.
   

# 2. Vor- und Nachteile / Pros and cons (German only)

im Vergleich zu papierbasierter Erfassung der Anwesenheit

Vorteile:
- Zeitersparnis Erfassung: 
  Die Erfassung geschieht mit viel weniger Zeit- und Arbeitsaufwand.
- Zeitersparnis Kontaktverfolgung:
  Im Infektionsfall kann dem Gesundheitsamt innerhalb von
  Minuten eine vollständige _elektronische_ Liste der möglichen
  Kontaktpersonen übermittelt werden.
- Ressourcenersparnis: 
  Spart erhebliche Mengen an Papier ein.
- Hygiene: 
  Es müssen keine Papiere von Hand zu Hand weitergegeben werden.
- Datenschutz 1: 
  Nur eine Person hat anlasslos Zugang zu den erfassten Daten.
  Im Papierfall sind dies viele und es ist schwierig zu überschauen, welche.
- Datenschutz 2:
  Die Daten können am Ende der Aufbewahrungsfrist vollautomatisch und
  somit verlässlich gelöscht werden.

Risiken:
- Datenschutz:
  Ein gegebenenfalls erfolgreicher Einbruch in die Datenbank des Servers
  beträfe mehr Daten als
  ein Diebstahl papierner Erfassungsbögen.  
  (Das Risiko ist klein, denn die Datenbank ist für Angreifer nicht attraktiv,
   weil sie keine wertvollen Daten wie Bankdaten oder Passwörter enthält.)
- Vollständigkeit:
  Der Erfassungsvorgang ist für die anderen Anwesenden weniger anschaulich
  und kann deshalb möglicherweise leichter vergessen werden. 
- Richtigkeit:
  Unsinnige Eingaben von Besuchern ("Donald Duck") sind kaum mehr 
  an Ort und Stelle zu entdecken.


# 3. Der/Die Datenverwalter/in / The main administrator role (German only)

- Benötigt eine Einweisung (ca. 1 Stunde).
- Muss Mitglied einer teilnehmenden Hochschule sein.
  Alle anderen teilnehmenden Hochschulen müssen mit dieser Hochschule
  einen Auftragsdatenverarbeitungsvertrag schließen.
- Sollte bei jeder Meldung von Räumen aus einer Hochschuleinheit
  die Excel-Datei durchsehen und nötigenfalls die Schreibweisen
  (insbesondere von Hochschuleinheiten) vereinheitlichen.
- Muss bei jeder Meldung von Räumen aus einer Hochschuleinheit klären,
  wer für diese Einheit berechtigt ist, Anwesenheitsdaten abzurufen,
  und dann den Datenzugang auf diesen Personenkreis beschränken.
  

# 4. Deployment and operation

Everything beyond this point is technical information, therefore in English.

The application is meant to be deployed locally in most organizations
(to simplify the situation regarding privacy protection)
and allows some configuration to adapt to local needs.  
It is designed such that one installation can serve several smaller
organizations together, though.


## 4.1 Architecture

The service is build using Python, Django, PostgreSQL, Gunicorn, and Traefik.
The deployment procedure described below will obtain these pieces
and configure them.
The code organization roughly follows the
[cookiecutter-django](https://cookiecutter-django.readthedocs.io) template
and the Python code is structured Django-style.


## 4.2 Deployment overview

(If you find errors anywhere, please speak up!)

The deployment procedure assumes an existing infrastructure of
Linux, bash, ssh, rsync, and Docker 18.09 or younger 
(with docker-compose 1.21 or younger).
The deploying user must be a member of the `docker` group.

There are six possible styles of deployment:
- `DEPLOYMODE=CERTS`: a stand-alone configuration that brings its own Traefik web server
  and relies on a manually created certificate for https.
  This configuration uses three docker containers: traefik, django, postgres.
- `DEPLOYMODE=LETSENCRYPT`: a variant of the above that
  relies on Let's Encrypt to generate the certificates for https.
- `DEPLOYMODE=GUNICORN`: a configuration without webserver that exposes the
  service port of the Gunicorn application server to be used by 
  an existing webserver that is capable of https. 
  This configuration uses only two docker containers: django, postgres.

All three of these can be deployed locally on a single server (`REMOTE=0`)
or have their docker images build on a local machine
and deployed onto a remote server via a docker registry 
(`REMOTE=1`, resulting in another three styles of the same three DEPLOYMODEs).
In any case, there are only few manual steps; most work is done by a 
few calls to a script called `anw.sh`.

(If you want to deploy without docker, see
[GitHub Issue 10](https://github.com/prechelt/anwesende/issues/10)
for hints.)


## 4.3 Deployment procedure

1. Create a working directory anywhere on a Linux machine 
   (called the 'source machine').
   If your server is in a DMZ and cannot connect to other servers itself,
   this must be a separate machine (`REMOTE=1`). 
   Perform
   `git clone https://git.imp.fu-berlin.de/anwesende/anwesende.git`.
   Rename the directory if you want, e.g.: `mv anwesende anw`,  
   and go there: `cd anw`.  
   This working directory is the reference for all subsequent commands.
2. Do `./anw.sh - prepare_envs`.
   This creates two new files:
   - `.envs/myenv.env` is a docker environment file containing various
     application settings. Carefully review the comments in that file and insert
     appropriate values for each setting.
   - `.envs/production.sh` is a shell environment file containing various
     deployment settings. Carefully set appropriate values here, too.
3. If `REMOTE=1`, call `./anw.sh production docker_login` and type
   your password twice to log into the docker registry locally and on the server.
   (If `REMOTE=0`, skip this step.)
4. Do `./anw.sh production install`.
   If all goes well, this command will
   - build the docker images on the source machine
   - if `REMOTE=1`, transfer them to the server's docker service via the registry
     and copy a handful of configuration files to the server's `~/anw` directory. 
   - create and start the docker containers on the server.
   Carefully review the output for error messages. 
   Repair those problems and repeat the command.
5. For `DEPLOYMODE=CERTS` only:  

RELEASES.md
# a.nwesen.de Releases

Release dates, version numbers, and contents:


## Versions 0.x

- 2020-10-06, Version **0.1**: 
  - Started development: Use case descriptions
- 2020-10-21, Version 0.6: 
  - Basic functionality is complete:
    Login, reading Excel files, generating QR codes, visitor registration form,
    retrieving visit groups and writing Excel file
- 2020-10-28, Versison 0.7: 
  - User-visible process documentation
  - Demo mode
  - Automatic purging of visit data after retention time
- 2020-11-06, Version 0.8: 
  - Pilot deployment, deployment description
  - Load testing (about 1000 visits/minute: fast enough)
  - Logging


## Version 1.0

- 2020-12-29, Version **1.0**: 
  - `anw.sh` install script for mostly automated deployment


## Versions 2.x

- 2021-01-02, Version **2.0**:
  - introduced 2-dimensional seat numbering and distance calculation
  - This is a semi-incompatible change: Existing pasted QR codes
    will show e.g. seat 14, but will in fact now refer to seat r1s14;
    existing importstep objects will lose their seats.
- 2021-01-05, Version 2.1:
  - added FAQ page
  - updated `/import` (which had outdated documentation in version 2.0)
  - more info on the QR code snippets: URL, instruction
  - published at GitHub
- 2021-01-06, Version 2.2:
  - README now also describes DB restore and deployment identification
- 2021-03-14, Version 2.3:
  - now shows number of people in room after one has registered
  - introduced `DEPLOYMODE=DEVELOPMENT` for development
    and removed some outdated deployment-related stuff 
  - fixed mistakes in `anw.sh` script, install instructions, FAQ.
  - updated Python to 3.9 and removed Pillow dependency 
- 2021-03-29, Version 2.4:
  - fixed and improved import statistics display
  - added a usage statistics (data status) table page:  
    Organization, Department, #rooms, #seats, #visits
  - fixed a timezone issue
  

## Versions 3.x

- 2021-04-23, Version **3.0**:
  - introduced setting COOKIE_WITH_RANDOMSTRING for optionally turning off
    the unique cookie identifier that is normally included in the data
  - introduced setting USE_EMAIL_FIELD for optionally excluding the
    email address from the registration form (because some states do
    not mention this in their Infektionsschutzverordnung).
    Note that not having email adresses _greatly_ reduces the value of
    running a.nwesen.de in the first place. Not recommended!
  - created space for smartphone keyboard below the registration form
    to accomodate some ill-behaving browsers
  - Page `/qrcodes/x` now formats vertically on smartphones.
- 2021-05-21, Version 3.1:
  - now properly converted times from UTC to localtime in the Excel output.
- 2021-06-06, Version 3.2:
  - added show_rooms
  - introduced `GUNICORN_WORKERS` and `GUNICORN_THREADS` settings.
    These settings increase performance, in particular for servers with
    multiple CPUs.
    Existing installations must add this in their `.envs/myenv.env` file; 
    see `config/envs/myenv-template.env` for how this will look like.
  - The search dialog now ignores case (valuable for email!).  
  - fixed failure when retrieving contact groups for empty search results
- 2021-06-15, Version 3.3:
  - made access control stricter 
    (the usage statistics and show-QR-codes pages now require a login
    in order to reduce the confusion when a Datenverwalter's session expires,
    which would previously result in crippled output)


## Versions 4.x

- 2021-07-12, Version 4.0:
  - added columns `row_dist` and `seat_dist` for the distance (in meters)
    between rows (front-to-back direction) and seats (left-to-right direction), 
    respectively, so that different rooms can have seating arrangements
    of different density and still the distances computed are reasonably
    accurate.
    For existing rooms, 1.4m is assumed for both values.  
    The new columns can be added to the database by reading an updated version
    of previous Excel files with identical values of
    `organization`, `department`, `building`, and `room`.
    See FAQ 3.3 for a discussion.    
- 2021-08-17, Version 4.1:
  - corrected wrong columname in some file import "Excel error" messages
  - improved logging
  - `/show_rooms` now displays `row_dist` and `seat_dist`
  - made Excel import robust against non-text cell contents
- 2021-09-20, Version 4.2:
  - Limit of 99 seats and rows replaced by 2000 seats total.
  - faq.html now discusses U-shaped seating arrangements (a common hurdle
    for the understanding of room admins).


## Versions 5.x

- 2021-09-25, Version 5.0:
  - introduce `status_3g` vaccination status field and 
    `USE_STATUS_3G_FIELD` feature toggle to use it (default: `True`)
  - added FAQ 3.4 "Wie informiere ich meine Besucher_innen?" including
    a proposed hang-on-the-wall information.
- 2021-09-28, Version 5.1:
  - greatly improved `loadtest.py`
  - server setup hardened: Gunicorn: add `--max-requests 4000`;
    the generated `docker-compose.yml` will now specify `restart: unless-stopped`
  - when not logged in, properly redirect upon `/import` upload (instead of just failing)  
- 2021-10-29, Version 5.2:
  - visit form retrieves vaccinated&recovered states from cookie, 
    but tested state not
  - show visit form errors very prominently
  - search "Besuche finden" shows date and time
  - updated django Docker base image to python:3.9-slim-bullseye
  - catch exception for invalid cookies
  - made docker build process a lot leaner and faster
- 2021-11-08, Version 5.3:
  - times of day can also be entered as hhmm, not only as hh:mm
  - define setting `DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'`
    to get rid of warnings from Django 3.2.
  - introduced setting `DATA_RETENTION_DAYS_STATUS_3G`
    and extended the `delete_outdated_data` management command
    to cleanse `status_3g` field values after that many days.  
    Just set this to the same value as `DATA_RETENTION_DAYS`
    if you do not need it.

anwesende/room/excel.py
"""
Excel import (for creating QR codes) and Excel export (for attenance lists).
Import also validates the proper file format.
Export also controls which groups of visitors to include.
"""
import collections
import os
import re
import tempfile
import typing as tg

import django.db.models.query as djdmq
from django.conf import settings

import anwesende.room.models as arm
import anwesende.users.models as aum
import anwesende.utils.date as aud
import anwesende.utils.excel as aue


class InvalidExcelError(ValueError):
    pass  # no additional logic is needed


def validate_excel_importfile(filename) -> None:
    # may raise InvalidExcelError
    columnsdict = aue.read_excel_as_columnsdict(filename)
    _validate_room_declarations(columnsdict)


def create_seats_from_excel(filename: str, user: aum.User) -> arm.Importstep:
    columnsdict = aue.read_excel_as_columnsdict(filename)
    _validate_room_declarations(columnsdict)
    importstep = _create_importstep(user)
    rooms, importstep.num_new_rooms, importstep.num_existing_rooms = \
        _find_or_create_rooms(columnsdict, importstep)
    seats, importstep.num_new_seats, importstep.num_existing_seats = \
        _find_or_create_seats(rooms)
    importstep.save()
    return importstep


def _validate_room_declarations(columndict: aue.Columnsdict):
    if getattr(columndict, 'has_been_validated', False): 
        return
    _validate_columnlist(columndict)
    _validate_single_department(columndict)
    _validate_dist(columndict, 'row_dist')
    _validate_dist(columndict, 'seat_dist')
    _validate_seatrange(columndict)
    columndict.has_been_validated = True  # type: ignore


def _validate_columnlist(columndict: aue.Columnsdict):
    expected = ('organization', 'department', 'building', 'room', 
                'row_dist', 'seat_dist', 'seat_last')
    found = columndict.keys()
    missing = set(expected) - set(found)
    if missing:
        _excelerror(row=1, expected=f"columns {expected}",
                    found=f"some columns are missing: {missing}")
    surprises = set(found) - set(expected)
    if surprises:
        _excelerror(row=1, expected=f"columns {expected}",
                    found=f"there are additional columns: {surprises}")


def _validate_single_department(columndict: aue.Columnsdict):
    organizations = set(columndict['organization'])
    departments = set(columndict['department'])
    if len(organizations) > 1:
        _excelerror(column='organization',
                    expected="all values are the same",
                    found=f"multiple different values: {organizations}")
    if len(departments) > 1:
        _excelerror(column='department',
                    expected="all values are the same",
                    found=f"multiple different values: {departments}")


def _validate_dist(columndict: aue.Columnsdict, colname: str) -> None:
    assert colname in ['row_dist', 'seat_dist']
    my_dists = columndict[colname]
    my_dist_format = r"\d[\.,]\d\d?"
    for index, my_dist in enumerate(my_dists):
        excel_row_number = index + 2
        try:
            if not re.fullmatch(my_dist_format, my_dist):
                format_msg = "distance value must look like '1.2' or '1,25'"
                _excelerror(row=excel_row_number, column=colname,
                            found=f"'{my_dist}': {format_msg}")
            else:
                my_dist = arm.Room.as_float(my_dist)
                if my_dist < 0.3 or my_dist > 4.0:
                    distance_msg = "distance must be in range 0.3 to 4.0 meters"
                    _excelerror(row=excel_row_number, column=colname,
                                found=f"'{my_dist}': {distance_msg}")
        except Exception as ex:
            if isinstance(ex, InvalidExcelError):
                raise ex
            else:
                _excelerror(row=excel_row_number, column=colname,
                            found=str(ex))


def _validate_seatrange(columndict: aue.Columnsdict):
    last_seats = columndict['seat_last']
    for index in range(len(last_seats)):
        excel_row_number = index + 2
        try:
            entry = last_seats[index]
            maxrow, maxseat = arm.Seat.split_seatname(entry)
            if maxrow*maxseat not in range(1, 2000):
                msg = f"{entry}: rooms must have 1 to 2000 seats, not {maxrow*maxseat}"
                _excelerror(row=excel_row_number, column='seat_last', found=msg)
        except ValueError as ex:
            _excelerror(row=excel_row_number, column='seat_last',
                        found=str(ex))


def _create_importstep(user: aum.User) -> arm.Importstep:
    result = arm.Importstep(user=user)
    result.save()
    return result


def _find_or_create_rooms(
        columnsdict: aue.Columnsdict,
        importstep: arm.Importstep) -> tg.Tuple[tg.Sequence[arm.Room], int, int]:
    _validate_room_declarations(columnsdict)
    result = []
    newN = existingN = 0
    for idx in range(len(columnsdict['room'])):
        def col(name): 
            return columnsdict[name][idx] 
        room, created = arm.Room.objects.get_or_create(
            organization=col('organization'),
            department=col('department'),
            building=col('building'),
            room=col('room'),
            defaults=dict(row_dist=arm.Room.as_float(col('row_dist')), 
                          seat_dist=arm.Room.as_float(col('seat_dist')),
                          seat_last=col('seat_last'),
                          importstep=importstep)
        )
        if created:
            newN += 1
        else:
            existingN += 1
            room.importstep = importstep
            #--- some fields can be modified by re-imports:
            room.row_dist = arm.Room.as_float(col('row_dist'))
            room.seat_dist = arm.Room.as_float(col('seat_dist'))
            room.seat_last = col('seat_last')
            room.save()
        result.append(room)
    return (result, newN, existingN)


def _find_or_create_seats(rooms: tg.Sequence[arm.Room]) \
        -> tg.Tuple[tg.Sequence[arm.Seat], int, int]:
    result = []
    newN = existingN = 0
    for room in rooms:
        maxrow, maxseat = arm.Seat.split_seatname(room.seat_last)
        for rownum in range(1, maxrow + 1):
            for seatnum in range(1, maxseat + 1):
                seat, created = arm.Seat.objects.get_or_create(
                    rownumber=rownum,
                    seatnumber=seatnum,
                    room=room,
                    defaults=dict(hash=arm.Seat.seathash(
                        room, arm.Seat.form_seatname(rownum, seatnum)))
                )
                result.append(seat)
                if created:
                    newN += 1
                else:
                    existingN += 1
    return (result, newN, existingN)


def _excelerror(row: int = None, column: str = None,
                expected: str = None, found: str = None
                ):
    assert found  # the core part, must not be left empty
    result = "Excel error: "
    if row:
        result += f"row {row}  "
    if column:
        result += f"column '{column}':"
    if row or column:
        result += '\n'
    if expected:
        result += f"EXPECTED: {expected}.\n"
    if expected:
        result += "FOUND: "
    result += found
    raise InvalidExcelError(result)


VGroupRow = collections.namedtuple('VGroupRow',  # noqa
    'familyname givenname email phone street_and_nr zipcode town '
    'status_3g '
    'cookie '
    'distance '
    'when from_time to_time '
    'organization department building room seat ')


Visits = tg.List[tg.Optional[arm.Visit]]


Expl = collections.namedtuple('Expl', ("Erklaerung", ))


explanations = [
    Expl("Jede Zeile im Tabellenblatt 'Daten' ist ein Besuch einer Person."),
    Expl("Die Spalten beschreiben die Person, dann die Zeit, dann den Ort."),
    Expl("" if settings.USE_EMAIL_FIELD
         else "(Spalte 'email' ist abgeschaltet.)"),
    Expl("Leere Zeilen trennen Gruppen von Personen. Die Mitglieder einer Gruppe "
         "sind einander laut ihren Angaben mindestens "
         f"{settings.MIN_OVERLAP_MINUTES} Minuten im gleichen Raum begegnet."),
    Expl(""),
    Expl("**cookie**: Die Daten enthalten meist viele Personen mehrfach."),
    Expl("Um diese Duplikate zu überblicken, kann man nach Spalte "
         "'cookie' sortieren."),
    Expl("Das selbe Cookie steht meist für die selbe Person."),
    Expl("Eine Person hat mehrere verschiedene Cookies, wenn sie "
         "das Cookie gelöscht oder mehrere Geräte oder Browser benutzt hat."),
    Expl("Zwei Personen können das selbe Cookie haben, wenn eine Person "
         "ihr Gerät verleiht."),
    Expl("Wenn viele Personen mit fragwürdigen Angaben das selbe Cookie "
         "haben, sind die Daten vielleicht Fantasie-Angaben."),
    Expl("" if settings.COOKIE_WITH_RANDOMSTRING
         else "(Die Funktion ist abgeschaltet)"),
    Expl("**when**: Spalte 'when' enthält den Zeitpunkt der Meldung und "),
    Expl("'from' und 'to' die vom Besucher manuell eingegebenen Werte "
         "für den Zeitraum."),
    Expl("**distance**: Abstand vom Sitzplatz der infizierten Person"),
    Expl(""),
]


def collect_visitgroups(primary_visits: djdmq.QuerySet
                        ) -> Visits:
    result = []
    visit_pks_seen = set()  # all contacts of primary visits
    primary_visit_pks_seen = set()  # only primary visits
    primary_visits = list(primary_visits)  # retrieve them only once
    primary_visit_pks_all = { pvisit.pk for pvisit in primary_visits }
    for pvisit in primary_visits:
        pvisit.distance = pvisit.seat.distance_in_m(pvisit.seat)  # add attr
        if pvisit.pk not in primary_visit_pks_seen:
            primary_visit_pks_seen.add(pvisit.pk)
        else:
            continue  # pvisit was already mentioned in a previous group
        visit_pks_seen.add(pvisit.pk)
        group = pvisit.get_overlapping_visits()
        for visit in group:
            visit.distance = visit.seat.distance_in_m(pvisit.seat)  # add attr
            must_not_be_suppressed = (visit.pk == pvisit.pk)
            is_new = visit.pk not in visit_pks_seen
            if is_new or must_not_be_suppressed:
                result.append(visit)
            if visit.pk in primary_visit_pks_all:  # make no separate group for visit.pk
                primary_visit_pks_seen.add(visit.pk)  
        result.append(None)  # empty row as separator
    if len(result) > 0:
        del result[-1]  # remove trailing empty row
    return result


def get_excel_download(visits: Visits) -> bytes:
    rows = _as_vgrouprows(visits)
    rowslists = dict(Daten=rows, Erklaerungen=explanations)
    with tempfile.NamedTemporaryFile(prefix="export_", suffix=".xlsx",
                                     delete=False) as fh:
        filename = fh.name  # file is deleted in 'finally' clause
    try:
        aue.write_excel_from_rowslists(filename, rowslists,  # type: ignore
                                       indexcolumn=True)
        with open(filename, 'rb') as file:
            excelbytes = file.read()  # slurp. Won't be very large.
    finally:
        os.unlink(filename)
    return excelbytes


def _as_vgrouprows(visits) -> tg.List[tg.Optional[VGroupRow]]:
    vgrouprows: tg.List[tg.Optional[VGroupRow]] = []
    v: arm.Visit
    for idx, v in enumerate(visits):
        if v is None:
            row = None
        else:
            distance = getattr(v, 'distance', None)
            distance = "%5.1fm" % v.distance if distance is not None else "?"
            row = VGroupRow(
                v.familyname, v.givenname, v.email, v.phone,
                v.street_and_number, v.zipcode, v.town, 
                v.status_3g_txt,
                v.cookie, distance,  # type: ignore
                aud.dtstring(v.submission_dt, time=True), 
                aud.dtstring(v.present_from_dt, date=False, time=True),
                aud.dtstring(v.present_to_dt, date=False, time=True),
                v.seat.room.organization, v.seat.room.department, 
                v.seat.room.building, v.seat.room.room,
                v.seat.seatname)
        vgrouprows.append(row)
    return vgrouprows
